### ISSUE TITLE
[RFC] All-to-all permute for Ulysses Parallel

### ISSUE THREAD

**Issue by kwen2501** [pytorch/pytorch#178066]
### 🚀 The feature, motivation and pitch

### What is Ulysses Parallel
UP is an optimization to attention modules in Transformers for handling **long sequences**, commonly seen in multi- or Omni- modal training or inference. Before entering the attention, each rank holds `1/P` of the sequence, but all the head states. For the attention to occur in parallel on different heads, it needs to transform into a state where: each rank holds the full sequence, but `1/P` of the head states. 

More details of UP can be found in this [DeepSpeed blog](https://github.com/deepspeedai/DeepSpeed/tree/master/blogs/deepspeed-ulysses).

UP is being used in:
- [vLLM-Omni](https://github.com/vllm-project/vllm-omni/issues/192)
- Hugging Face accelerate
- Megatron-DeepSpeed
- Arctic plugin to vLLM ("Shift Parallelism")

### What is the communication pattern
It is a permutation + all-to-all (can be understood as a "global permutation").

On the input side, we need to cut on the hidden dimension (dim 1), send chunk `i` i.e. `input[:, i, :]` to destination `i`, as its output chunk `i` on seq dimension (dim 0), i.e. `out[i, :]`.

[Example code pointer in vLLM-Omni](https://github.com/vllm-project/vllm-omni/blob/1bd1d9d618996e841781ee9e67bab1f371bc6344/vllm_omni/diffusion/distributed/comm.py#L16):
```
def all_to_all_4D(
    input: torch.tensor, scatter_idx: int = 2, gather_idx: int = 1, group=None, use_sync: bool = False
) -> torch.tensor
```

<img width="960" height="540" alt="Image" src="https://github.com/user-attachments/assets/5dad8252-37e5-4c48-b297-60b0070ce17a" />





### Alternatives

`dist.all_to_all_single` can only cut input data on dim 0. 
As a result, [today's implementation](https://github.com/vllm-project/vllm-omni/blob/1bd1d9d618996e841781ee9e67bab1f371bc6344/vllm_omni/diffusion/distributed/comm.py#L42-L44) always involves a permutation of the input tensor first:
```
        # transpose groups of heads with the seq-len parallel dimension, so that we can scatter them!
        # (bs, seqlen/P, hc, hs) -reshape-> (bs, seq_len/P, P, hc/P, hs) -transpose(0,2)-> (P, seq_len/P, bs, hc/P, hs)
        input_t = input.reshape(bs, shard_seqlen, seq_world_size, shard_hc, hs).transpose(0, 2).contiguous()

        if seq_world_size > 1:
            dist.all_to_all_single(output, input_t, group=group)

```

### Additional context

  Typical values:
  - bs=1 (long-sequence training typically uses bs=1)
  - num_total_heads=32, head_dim=128 (LLaMA-2 style)
  - p=8 (sequence parallel degree)
  - global_seq=65536 (64K), so local_seq=8192
  - dtype = bf16 (2 bytes)

  Tensor size (one of Q/K/V):
  8192 * 1 * 32 * 128 * 2 bytes = 64 MB

  ---
  The permute operation

  - Reshape: [8192, 1, 32, 128] → [8192, 1, 8, 4, 128]  (free, just metadata)
  - Permute (2,0,1,3,4): [8192, 1, 8, 4, 128] → [8, 8192, 1, 4, 128]
  - `.contiguous()` ← this is the actual cost: forces a full memory copy

  The permute itself is free — `.contiguous()` is what triggers the copy. It reads 64 MB and writes 64 MB = 128 MB HBM traffic.

  ---
  Key hardware specs                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                             
```  ┌──────────────────────────────────┬───────────┬───────────┐                                                                                                                                                                                                               
  │                                  │ A100 SXM  │ H100 SXM5 │
  ├──────────────────────────────────┼───────────┼───────────┤                                                                                                                                                                                                               
  │ HBM bandwidth                    │ 2.0 TB/s  │ 3.35 TB/s │
  ├──────────────────────────────────┼───────────┼───────────┤                                                                                                                                                                                                               
  │ NVLink version                   │ 3.0       │ 4.0       │                                                                                                                                                                                                               
  ├──────────────────────────────────┼───────────┼───────────┤
  │ NVLink bandwidth (per direction) │ ~300 GB/s │ ~450 GB/s │                                                                                                                                                                                                               
  ├──────────────────────────────────┼───────────┼───────────┤                                                                                                                                                                                                               
  │ HBM / NVLink ratio               │ 6.7×      │ 7.4×      │
  └──────────────────────────────────┴───────────┴───────────┘                                                                                                                                                                                                               
```                                                            
  Per Q/K/V tensor                                                                                                                                                                                                                                                           
                                                            
  Permute (128 MB HBM traffic, ~70% efficiency):                                                                                                                                                                                                                             
```                                                            
  ┌──────┬──────────────┬────────┐                                                                                                                                                                                                                                           
  │      │ Effective BW │  Time  │                          
  ├──────┼──────────────┼────────┤
  │ A100 │ ~1.40 TB/s   │ ~91 µs │
  ├──────┼──────────────┼────────┤
  │ H100 │ ~2.35 TB/s   │ ~54 µs │                                                                                                                                                                                                                                           
  └──────┴──────────────┴────────┘                                                                                                                                                                                                                                           
```                                                                                                                                                                                                                                                                             
  All-to-all (64 MB over NVLink unidirectional):                                                                                                                                                                                                                             
```                                                            
  ┌──────┬───────────┬─────────┐                                                                                                                                                                                                                                             
  │      │ Bandwidth │  Time   │                            
  ├──────┼───────────┼─────────┤
  │ A100 │ ~300 GB/s │ ~213 µs │
  ├──────┼───────────┼─────────┤
  │ H100 │ ~450 GB/s │ ~142 µs │                                                                                                                                                                                                                                             
  └──────┴───────────┴─────────┘
```                                                                                                                                                                                                                                                                             
  Combined Q+K+V                                                                                                                                                                                                                                                             
   
```  ┌──────┬────────────┬───────────────┬─────────────────────┐                                                                                                                                                                                                                
  │      │ Permute ×3 │ All-to-all ×3 │ Permute as % of A2A │
  ├──────┼────────────┼───────────────┼─────────────────────┤                                                                                                                                                                                                                
  │ A100 │ ~273 µs    │ ~639 µs       │ ~43%                │
  ├──────┼────────────┼───────────────┼─────────────────────┤
  │ H100 │ ~162 µs    │ ~426 µs       │ ~38%                │
  └──────┴────────────┴───────────────┴─────────────────────┘                                                                                                                                                                                                                
```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan

**Comment by RohitRathore1** [comment]
interesting
